include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -50,6 +51,142 @@ namespace gko {
 namespace solver {
 
 
+/**
+ * Give a initial guess about the input of the apply method.
+ */
+enum class initial_guess {
+    /**
+     * the input is zero
+     */
+    zero,
+    /**
+     * the input is right hand side
+     */
+    rhs,
+    /**
+     * the input is given
+     */
+    given
+};
+
+
+class MultigridState;
+
+
+/**
+ * ApplyGuess provides a way to give the input guess for apply function.
+ * All functionalities have the protected access specifier. It should only be
+ * used internally.
+ */
+class ApplyGuess {
+protected:
+    friend class MultigridState;
+
+    /**
+     * Modify the input vector x by the guess
+     *
+     * @param b  the right hand side vectors
+     * @param x  the input vectors
+     * @param guess  the input guess
+     */
+    static void modify_input(const LinOp* b, LinOp* x, initial_guess guess)
```

Comment:
Move to source? This is not a class template. Also naming suggestion:
```suggestion
    static void prepare_initial_guess(const LinOp* b, LinOp* x, initial_guess guess)
```